File: courseMaterial/course_ROSActions/code_section_5_actions_activity/my_robot_tutorials/src/activity_action_server_queue.cpp
#include <ros/ros.h>
#include <actionlib/server/action_server.h>
#include <actionlib/server/server_goal_handle.h>

#include <thread>
#include <vector>
#include <map>
#include <string>
#include <queue>

#include <my_robot_msgs/CountUntilAction.h>
#include <my_robot_msgs/CountUntilGoal.h>
#include <my_robot_msgs/CountUntilResult.h>
#include <my_robot_msgs/CountUntilFeedback.h>

typedef actionlib::ServerGoalHandle<my_robot_msgs::CountUntilAction> GoalHandle;

class CountUntilServer {

    protected:
    ros::NodeHandle _nh;
    actionlib::ActionServer<my_robot_msgs::CountUntilAction> _as;

    std::vector<std::thread*> _goalThreads;
    std::map<std::string, bool> _cancelGoals;

    std::queue<GoalHandle> _goalQueue;
    std::thread* _goalQueueThread;

    public:
    CountUntilServer():
        _as(_nh, "/count_until",
            boost::bind(&CountUntilServer::onGoal, this, _1),
            boost::bind(&CountUntilServer::onCancel, this, _1),
            false)
    {
        _as.start();
        ROS_INFO("Action Server has been started.");

        _goalQueueThread = new std::thread(boost::bind(&CountUntilServer::runQueue, this));
    }

    ~CountUntilServer()
    {
        _goalQueueThread->join();
        delete _goalQueueThread;
    }

    void runQueue()
    {
        ros::Rate rate(2.0);
        while (ros::ok())
        {
            rate.sleep();
            if (_goalQueue.size() > 0) {
                processGoal(_goalQueue.front());
                _goalQueue.pop();
            }
            else {
                ROS_INFO("Not doing anything");
            }
        }
    }

    void processGoal(GoalHandle gh)
    {
        boost::shared_ptr<const my_robot_msgs::CountUntilGoal> goal = gh.getGoal();
        std::string goal_id = gh.getGoalID().id;
        int max_number = goal->max_number;
        double wait_duration = goal->wait_duration;
        ROS_INFO("Processing goal: %s", goal_id.c_str());
        ROS_INFO("Max number: %d, wait duration: %lf",
            max_number, wait_duration);

        // Validate parameters
        if (max_number > 10) {
            gh.setRejected();
            return;
        }
        gh.setAccepted();

        int counter = 0;
        ros::Rate rate(1.0/wait_duration);
        bool success = false;
        bool preempted = false;

        while (ros::ok())
        {
            counter++;
            ROS_INFO("%s: %d", goal_id.c_str(), counter);
            if (_cancelGoals[goal_id]) {
                preempted = true;
                break;
            }
            if (counter >= max_number) {
                success = true;
                break;
            }
            my_robot_msgs::CountUntilFeedback feedback;
            feedback.percentage = 
                (double)counter / (double)max_number;
            gh.publishFeedback(feedback);
            rate.sleep();
        }
        
        my_robot_msgs::CountUntilResult result;
        result.count = counter;
        ROS_INFO("Send goal result to client: %s", goal_id.c_str());

        if (preempted) {
            gh.setCanceled(result);
        }
        else if (success) {
            gh.setSucceeded(result);
        }
        else {
            gh.setAborted(result);
        }

        _cancelGoals.erase(goal_id);
    }

    void onGoal(GoalHandle gh)
    {
        ROS_INFO("Received new goal");
        
        _goalQueue.push(gh);
        //_goalThreads.push_back(
        //    new std::thread(boost::bind(&CountUntilServer::processGoal, this, gh)));

        _cancelGoals[gh.getGoalID().id] = false;
    }

    void onCancel(GoalHandle gh)
    {
        ROS_INFO("Cancel request has been received.");
        if (_cancelGoals.count(gh.getGoalID().id) > 0) {
            ROS_INFO("Found goal to cancel");
            _cancelGoals[gh.getGoalID().id] = true;
        }
    }

};

int main (int argc, char **argv)
{
    ros::init(argc, argv, "count_until_server");

    CountUntilServer server;
    ros::spin();
}

File: courseMaterial/course_ROSActions/code_section_5_actions_activity/my_robot_tutorials/scripts/activity_action_server_queue.py
#!/usr/bin/env python

import threading
import rospy
import actionlib

from my_robot_msgs.msg import CountUntilAction
from my_robot_msgs.msg import CountUntilGoal
from my_robot_msgs.msg import CountUntilResult
from my_robot_msgs.msg import CountUntilFeedback

class CountUntilServer:

    def __init__(self):
        self._as = actionlib.ActionServer('/count_until',
            CountUntilAction, self.on_goal, self.on_cancel,
            auto_start=False)
        self._as.start()
        rospy.loginfo("Action Server has been started.")
        self._cancel_goals = {}

        self._goal_queue = []
        w = threading.Thread(name="queue_worker", 
            target=self.run_queue)
        w.start()
        rospy.loginfo("Run queue thread has been started")

    def run_queue(self):
        rate = rospy.Rate(2.0)
        while not rospy.is_shutdown():
            rate.sleep()
            if len(self._goal_queue) > 0:
                self.process_goal(self._goal_queue.pop(0))
            else:
                rospy.loginfo("Not doing anything")

    def process_goal(self, goal_handle):
        goal_id = goal_handle.get_goal_id()
        rospy.loginfo("Processing goal : " + str(goal_id.id))
        goal = goal_handle.get_goal()
        max_number = goal.max_number
        wait_duration = goal.wait_duration

        # Validate parameters
        if max_number > 10:
            goal_handle.set_rejected()
            return
        goal_handle.set_accepted()

        counter = 0
        rate = rospy.Rate(1.0/wait_duration)

        success = False
        preempted = False

        while not rospy.is_shutdown():
            counter += 1
            rospy.loginfo(str(goal_id.id) + ":  " + str(counter))
            if self._cancel_goals[goal_id]:
                preempted = True
                break
            if counter >= max_number:
                success = True
                break
            feedback = CountUntilFeedback()
            feedback.percentage = \
                float(counter) / float(max_number)
            goal_handle.publish_feedback(feedback)
            rate.sleep()

        result = CountUntilResult()
        result.count = counter
        rospy.loginfo("Send goal result to client")

        if preempted:
            rospy.loginfo("Preempted")
            goal_handle.set_canceled(result)
        elif success:
            rospy.loginfo("Succeeded")
            goal_handle.set_succeeded(result)
        else:
            rospy.loginfo("Aborted")
            goal_handle.set_aborted(result)
        rospy.loginfo("--- Finished to process the goal : " + \
            str(goal_id.id))
        self._cancel_goals.pop(goal_id)

    def on_goal(self, goal_handle):
        rospy.loginfo("Received new goal")
        rospy.loginfo(goal_handle.get_goal())

        self._goal_queue.append(goal_handle)
        self._cancel_goals[goal_handle.get_goal_id()] = False
    
    def on_cancel(self, goal_handle):
        rospy.loginfo("Received cancel request.")
        goal_id = goal_handle.get_goal_id()

        if goal_id in self._cancel_goals:
            rospy.loginfo("Found goal to cancel")
            self._cancel_goals[goal_id] = True


if __name__ == '__main__':
    rospy.init_node('count_until_server')

    server = CountUntilServer()

    rospy.spin()
